FWCore/Framework/interface/CallbackBase.h

```diff
@@ -0,0 +1,270 @@
+              });
+            } catch (cms::Exception& caughtException) {
+              if (not excptr) {
+                exceptionContext(caughtException, callingContext_);
+                excptr = std::current_exception();
+              }
+            }
+          }
+          if (excptr) {
+            taskList_.doneWaiting(excptr);
+            return;
+          }
+
+          producer_->queue().push(*group, [this, serviceToken, record, eventSetupImpl]() {
+            callingContext_.setState(ESModuleCallingContext::State::kRunning);
+            std::exception_ptr exceptPtr;
+            try {
+              convertException::wrap([this, &serviceToken, &record, &eventSetupImpl] {
+                ESModuleCallingContext const& context = callingContext_;
+                auto proxies = getTokenIndices();
+                if (postMayGetProxies_) {
+                  proxies = &((*postMayGetProxies_).front());
+                }
+                TRecord rec;
+                ESParentContext pc{&context};
+                rec.setImpl(record, transitionID(), proxies, eventSetupImpl, &pc);
+                ServiceRegistry::Operate operate(serviceToken.lock());
+                record->activityRegistry()->preESModuleSignal_.emit(record->key(), context);
+                struct EndGuard {
+                  EndGuard(EventSetupRecordImpl const* iRecord, ESModuleCallingContext const& iContext)
+                      : record_{iRecord}, context_{iContext} {}
+                  ~EndGuard() { record_->activityRegistry()->postESModuleSignal_.emit(record_->key(), context_); }
+                  EventSetupRecordImpl const* record_;
+                  ESModuleCallingContext const& context_;
+                };
+                EndGuard guard(record, context);
+                decorator_.pre(rec);
+                storeReturnedValues((*produceFunction_)(rec));
+                decorator_.post(rec);
+              });
+            } catch (cms::Exception& iException) {
+              exceptionContext(iException, callingContext_);
+              exceptPtr = std::current_exception();
+            }
+            taskList_.doneWaiting(exceptPtr);
+          });
+        });
+      }
+
+      void prefetchAsync(WaitingTaskHolder iTask,
+                         EventSetupRecordImpl const* iRecord,
+                         EventSetupImpl const* iEventSetupImpl,
+                         ServiceToken const& token,
+                         ESParentContext const& iParent) {
+        bool expected = false;
+        auto doPrefetch = wasCalledForThisRecord_.compare_exchange_strong(expected, true);
+        taskList_.add(iTask);
+        if (doPrefetch) {
+          auto group = iTask.group();
+          ServiceWeakToken weakToken(token);
+          WaitingTaskHolder runModuleTaskHolder = makeRunModuleTaskHolder(group, weakToken, iRecord, iEventSetupImpl);
```

Comment:
@makortel I tried the Curiously Recurring Template Pattern. It's the third commit. A little extra complexity. Possibly premature optimization. I'm happy with it though. It was only a little extra complexity and was interesting to implement... It is the third commit if you want to take a look. I just pushed it.

Reply:
I agree CRTP is possibly a premature optimization. My main motivation was that the related code was already always dealing with the concrete `Callback` type without any virtual functions, and I wanted to preserve that (as long as it is feasible).

Reply:
I pushed the third commit right away because I wanted some feedback on that one. Do you want me to back the CRTP version out and go back to a using a virtual function? Both ways work and it is a trade off (a virtual function is slightly simpler for a human code reader, CRTP probably runs slightly faster). I'm OK either way, just let me know which you prefer.

(As for the rest, I was still working on revisions for the comments. I'll push new commits soon. I got sidetracked on the office move yesterday and then my home WiFi breaking this morning. I spent half the morning talking to COMCAST... It is back up. I hope it stays that way.)

Reply:
I have a 3rd suggestion which avoids the virtual function (which I'd prefer we do not use) and CRTP. You could create a protected function in the base class that takes a templated pointer to the actual derived type,
```
template<typename DERIVED>
WaitingTaskHolder prefetchAsyncImpl(DERIVED* derived, ....)
```

Then require inheriting classes define the regular `prefetchAsync` which just calls `prefetchAsyncImpl` with the `this` pointer added. We do something similar with WorkerT and Worker.

Reply:
Hmm, if the only thing it needs is the value returned from `makeRunModuleTaskHolderImpl` then maybe it would be even simpler just to have

```
template<typename RunModuleFnctr>
WaitingTaskHolder prefetchAsyncImpl(RunModuleFnctr&&, ...)
```

and then prefetchAsync is just
```
WaitingTaskHolder prefetchAsync(...) {
   return prefetchAsyncImpl( [this](auto&& group, auto&& token, auto&& record, auto&& es) { return makeRunModuleTaskHolder(group, token, record, es);}, ...);
```